ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1079 @@
+					     uint8_t *fw_img, size_t fw_img_sz,
+					     paddr_t *rsc_pa,
+					     size_t *rsc_size)
+{
+	uint32_t da = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	Elf32_Word size = 0;
+
+	res = e32_parser_find_rsc_table(fw_img, fw_img_sz, &da, &size);
+	if (res)
+		return res;
+
+	if (da) {
+		DMSG("Resource table device address %#"PRIx32" size %#"PRIx32,
+		     da, size);
+
+		*rsc_pa = remoteproc_da_to_pa(da, size, ctx);
+		if (!*rsc_pa)
+			return TEE_ERROR_ACCESS_DENIED;
+	}
+
+	*rsc_size = size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_hash_table(struct remoteproc_context *ctx)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *tlv_hash = NULL;
+	struct remoteproc_segment *hash_table = NULL;
+	size_t length = 0;
+
+	/* Get the segment's hash table from TLV data */
+	res = remoteproc_get_tlv(ctx->tlvs, ctx->tlvs_sz, RPROC_TLV_HASHTABLE,
+				 &tlv_hash, &length);
+	if (res || (length % sizeof(struct remoteproc_segment)))
+		return res;
+
+	/* We can not ensure that tlv_hash is memory aligned so make a copy */
+	hash_table = TEE_Malloc(length, TEE_MALLOC_FILL_ZERO);
+	if (!hash_table)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	memcpy(hash_table, tlv_hash, length);
+
+	ctx->hash_table = hash_table;
+	ctx->nb_segment = length / sizeof(struct remoteproc_segment);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_tlv_images_type(struct remoteproc_context *ctx,
+				      uint8_t num_img, uint8_t idx,
+				      uint8_t *img_type)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *tlv_value = NULL;
+	size_t length = 0;
+
+	/* Get the type of the image to load, from TLV data */
+	res = remoteproc_get_tlv(ctx->tlvs, ctx->tlvs_sz, RPROC_TLV_IMGTYPE,
+				 &tlv_value, &length);
+	if (res || length != (sizeof(*img_type) * num_img))
+		return res;
+
+	*img_type = tlv_value[idx];
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_tlv_images_size(struct remoteproc_context *ctx,
+				      uint8_t num_img, uint8_t idx,
+				      uint32_t *img_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *tlv_value = NULL;
+	size_t length = 0;
+
+	/* Get the size of the image to load, from TLV data */
+	res = remoteproc_get_tlv(ctx->tlvs, ctx->tlvs_sz, RPROC_TLV_IMGSIZE,
+				 &tlv_value, &length);
+	if (res || length != (sizeof(*img_size) * num_img))
+		return res;
+
+	*img_size = LE32_TO_CPU(&tlv_value[sizeof(*img_size) * idx]);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_segment_hash(struct remoteproc_context *ctx, uint8_t *src,
+				   uint32_t size, uint32_t da,
+				   uint32_t mem_size, uint8_t **hash)
+{
+	struct remoteproc_segment *peh = NULL;
+	unsigned int i = 0;
+	unsigned int nb_entry = ctx->nb_segment;
+
+	peh = (void *)(ctx->hash_table);
+
+	for (i = 0; i < nb_entry; peh++, i++) {
+		if (peh->phdr.p_paddr != da)
+			continue;
+
+		/*
+		 * Segment is read from a non secure memory. Crosscheck it using
+		 * the hash table to verify that the segment has not been
+		 * corrupted.
+		 */
```

Comment:
I think this inline comment is misplaced or misphrased. It explains why we need a hash table, which rather belongs to `remoteproc_load_segment()` function, right below.
I think here the inline comment could state that segment info are crosschecked (load addres, load size, segment size, ...) against same info saved in hash table segment (read here from secure memory).

Suggestion (IIUC):
```
		/*
		 * Segment metadata were read from a non-secure memory.
                 * Validate them using hash table data stored in secure memory.
		 */